File: 22-2.py
import re
import sys
import json
import itertools
from collections import defaultdict
sys.setrecursionlimit(5000)

def deal(stack):
    return list(reversed(stack))

def deal_p(c, lenstack):
    return lenstack - c - 1

def cut(stack, i):
    if i > 0:
        return stack[i:] + stack[0:i]
    else:
        return stack[len(stack) + i:] + stack[0:len(stack) + i]

def cut_p(c, i, lenstack):
    if i < 0:
        i = lenstack + i
    if c < i:
        return c + (lenstack - i)
    else:
        return c - i
    
def incr(stack, i):
    newstack = stack.copy()
    p = 0
    for s in range(0, len(stack)):
        newstack[p] = stack[s]
        p += i
        if p >= len(stack):
            p -= len(stack)
    return newstack

def incr_p(c, i, lenstack):
    pos = c * i
    mod = pos % lenstack
    return mod
    
# modpow the polynomial: (ax+b)^m % n
# f(x) = ax+b
# g(x) = cx+d
# f^2(x) = a(ax+b)+b = aax + ab+b
# f(g(x)) = a(cx+d)+b = acx + ad+b
def polypow(a,b,m,n):
    if m==0:
        return 1,0
    if m%2==0:
        return polypow(a*a%n, (a*b+b)%n, m//2, n)
    else:
        c,d = polypow(a,b,m-1,n)
        return a*c%n, (a*d+b)%n

tot = 101741582076661
lenstack = 119315717514047
card = 2020

with open('22.txt') as f:
    ops = [l.strip() for l in f.readlines()]

    #https://stackoverflow.com/questions/4798654/modular-multiplicative-inverse-function-in-python
    # modinv(x, p)
    # y = pow(x, p-2, p)

    a = 1
    b = 0
    for x in reversed(ops):
        if x == 'deal into new stack':
            #print('deal();')
            #ops.append(('DEAL', 0))
            a = -a
            b = lenstack - b - 1
        elif x.startswith('deal with increment'):
            #print('incr({});'.format(int(x[len('deal with increment '):])))
            #ops.append(('INCR', int(x[len('deal with increment '):])))
            i = int(x[len('deal with increment '):])
            z = pow(i, lenstack - 2, lenstack)
            a = (a*z) % lenstack
            b = (b*z) % lenstack
        else:
            #print('cut({});'.format(int(x[len('cut '):])))
            #ops.append(('CUT', int(x[len('cut '):])))
            c = int(x[len('cut '):])
            b = (b+c) % lenstack

    print(a,b)
    #print( pow(a*2020 + b, 2, lenstack) )
    a,b = polypow(a,b, tot, lenstack)

    print('ans=', (card*a+b)%lenstack)

    lenstack = 10007
    a = 1
    b = 0
    for x in reversed(ops):
        if x == 'deal into new stack':
            #print('deal();')
            #ops.append(('DEAL', 0))
            a = -a
            b = lenstack - b - 1
        elif x.startswith('deal with increment'):
            #print('incr({});'.format(int(x[len('deal with increment '):])))
            #ops.append(('INCR', int(x[len('deal with increment '):])))
            i = int(x[len('deal with increment '):])
            z = pow(i, lenstack - 2, lenstack)
            a = (a*z) % lenstack
            b = (b*z) % lenstack
        else:
            #print('cut({});'.format(int(x[len('cut '):])))
            #ops.append(('CUT', int(x[len('cut '):])))
            c = int(x[len('cut '):])
            b = (b+c) % lenstack
    print((a*2019+b) % lenstack)
    a,b = polypow(a,b, 1, 10007)
    print(a,b)
    print('p1=', (1538*a+b)%lenstack)

    exit()

    for c in range(0, 3):
        print('calc', card)
        for op in ops:
            if op[0] == 'DEAL':
                card = deal_p(card, lenstack)
            elif op[0] == 'CUT':
                card = cut_p(card, op[1], lenstack)
            else:
                card = incr_p(card, op[1], lenstack)

    exit()

    print(card)

File: 22-2.c
#include <stdio.h>
#include <inttypes.h>

#define uint128_t unsigned __int128
#define int128_t __int128

uint128_t card = 2020;
uint128_t lenstack = 119315717514047;
uint128_t tot = 101741582076661;

/*
uint128_t card = 2019;
uint128_t lenstack = 10007;
uint128_t tot = 1;
*/

void deal()
{
    card = lenstack - card - 1LL;
}

void cut(int128_t i)
{
    if (i < 0)
        i = lenstack + i;
    if (card < i)
        card += (lenstack - i);
    else
        card -= i;
}

void incr(int i)
{
    card = (card * (uint128_t)i) % lenstack;
}

int main()
{
    for (uint128_t i = 0; i < tot; i++)
    {
        cut(4913);
        incr(53);
        cut(-1034);
        deal();
        incr(49);
        deal();
        cut(-7153);
        incr(59);
        cut(4618);
        incr(15);
        cut(-8047);
        incr(67);
        cut(9890);
        incr(16);
        cut(-870);
        incr(34);
        cut(-4557);
        incr(19);
        cut(-6466);
        incr(47);
        cut(-5860);
        incr(65);
        deal();
        cut(-8104);
        incr(15);
        cut(-9013);
        deal();
        cut(7309);
        incr(36);
        deal();
        cut(-1340);
        incr(42);
        cut(-5204);
        incr(75);
        deal();
        incr(16);
        deal();
        incr(44);
        cut(6833);
        incr(14);
        deal();
        cut(2345);
        incr(60);
        cut(4830);
        incr(75);
        cut(-2843);
        incr(50);
        cut(3816);
        deal();
        cut(7340);
        incr(48);
        cut(-3452);
        incr(62);
        cut(-2433);
        incr(59);
        cut(-4176);
        deal();
        cut(9365);
        incr(65);
        cut(-7815);
        incr(65);
        cut(5177);
        deal();
        incr(8);
        cut(-3200);
        incr(63);
        cut(-9460);
        incr(20);
        cut(-6926);
        incr(4);
        cut(-9863);
        incr(38);
        cut(-4295);
        incr(31);
        deal();
        cut(-9590);
        incr(22);
        cut(-9805);
        incr(48);
        deal();
        cut(-2326);
        incr(72);
        deal();
        incr(7);
        cut(-453);
        incr(15);
        cut(-7639);
        deal();
        incr(61);
        cut(-7303);
        incr(24);
        cut(4827);
        incr(48);
        cut(-1821);
        incr(31);
        cut(-9410);
        incr(5);
        cut(-2492);
        incr(25);
        cut(1313);

        if (i > 0 && i % 10000000 == 0)
        {
            printf("at %" PRIu64 "  %f\n", (uint64_t)i, i / tot);
            fflush(stdout);
            return 0;
        }
    }

    printf("card = %" PRIu64 "\n", (uint64_t)card);
}